servers/parser-graphql-wrapper/src/models/PocketMetadataModel.ts

```diff
@@ -0,0 +1,177 @@
+
+export interface IPocketMetadataDataSource {
+  matcher: RegExp;
+  ttl: number; // The ttl of the data in seconds
+  source: PocketMetadataSource;
+  version: number; // the version of the source parser
+  derivePocketMetadata(
+    item: Item,
+    fallbackParserPocketMetadata: PocketMetadata,
+    context: IContext,
+  ): Promise<PocketMetadata>;
+  isEnabled(context: IContext): boolean;
+}
+
+export class PocketMetadataModel {
+  constructor(
+    private readonly db: IPocketMetadataDataStore,
+    private readonly datasources: IPocketMetadataDataSource[],
+  ) {}
+
+  public async derivePocketMetadata(
+    item: Item,
+    context: IContext,
+    refresh: boolean,
+  ): Promise<PocketMetadata> {
+    const url = item.givenUrl; // the url we are going to key everything on.
+    const fallbackParserPocketMetadata: ItemSummary = {
+      id: item.id,
+      image: item.topImage ?? item.images?.[0],
+      excerpt: item.excerpt,
+      title: item.title ?? item.givenUrl,
+      authors: item.authors,
+      domain: item.domainMetadata,
+      datePublished: item.datePublished
+        ? DateTime.fromSQL(item.datePublished, {
+            zone: config.mysql.tz,
+          }).toJSDate()
+        : null,
+      url: url,
+      source: PocketMetadataSource.PocketParser,
+      __typename: 'ItemSummary',
+    };
+
+    // First we filter to our sources.
+    // We do this first because some sources could be behind a feature flag or not enabled
+    // We also only store other data sources beyond our parser in the datastore, \
+    // since the parser is cached elsewhere in Pocket
+    const sources = this.datasources.filter((datasource) => {
+      const pass =
+        datasource.isEnabled(context) && datasource.matcher.test(url);
+      return pass;
+    });
+    if (sources.length == 0) return fallbackParserPocketMetadata;
+
+    const source = sources[0];
+
+    if (!refresh) {
+      const storedSummary = await this.getPocketMetadata(
+        url,
+        source.version,
+        source.source,
+      );
+      // we need to ensure the stored source we pulled out of the dynamodb matches the one we are looking
+      // for in case 2 sites are used by multipe sources or an older less specific source
+      // TODO: In the future we should probably do a more specific select of the dynamodb data
+      if (storedSummary && source.source == storedSummary.source) {
+        return { ...storedSummary, item };
+      }
+    }
+
+    const newPocketMetadata = await sources[0].derivePocketMetadata(
+      item,
+      fallbackParserPocketMetadata,
+      context,
+    );
+    if (newPocketMetadata == null) return fallbackParserPocketMetadata;
+
+    // specifically we do not await this, so its a non-blocking call.
+    await this.savePocketMetadata(
+      newPocketMetadata,
+      source.ttl,
+      source.version,
+    );
+
+    return { ...newPocketMetadata, item };
+  }
+
+  /**
+   * Convert input to DynamoDB entity
+   */
+  toEntity(
```

Comment:
I think ultimately these are going to be implemented by subclasses but this is fine for now